File: src/lib/features/library/contracts.ts
export interface PlatformIdentity {
  source: string;
  id: string;
}

export interface GameIdentity {
  launchPath: string | null;
  platformId: PlatformIdentity | null;
  titleFingerprint: string;
}

export interface ImportSourceRecord {
  sourceRecordId: string;
  title: string;
  launchPath: string | null;
  installDir: string | null;
  platformId: PlatformIdentity | null;
  launchUri: string | null;
  fields: Record<string, unknown>;
}

export type ImportAction = "create" | "update" | "merge" | "conflict" | "ignore";
export type ImportReasonCode =
  | "new_identity"
  | "launch_path_match"
  | "platform_id_match"
  | "ambiguous_strong_identity"
  | "title_recall_only"
  | "no_launch_target"
  | "explicit_merge"
  | "explicit_ignore"
  | "stale_preview"
  | "invalid_decision";

export interface ImportReason {
  code: ImportReasonCode;
  message: string;
  recalledGameIds: string[];
}

export type IdentityMatchKind = "platform_id" | "launch_path" | "title_recall";

export interface IdentityMatch {
  gameId: string;
  gameTitle: string;
  kind: IdentityMatchKind;
  confidence: number;
}

export type FieldDiffDisposition =
  | "unchanged"
  | "fill_empty"
  | "replace_imported"
  | "preserve_user"
  | "conflict";

export interface FieldProvenance {
  gameId: string;
  field: string;
  source: string;
  sourceRecordId: string;
  importedAt: string;
  appliedValue: unknown;
  valueHash: string;
}

export interface FieldDiff {
  field: string;
  current: unknown;
  incoming: unknown;
  disposition: FieldDiffDisposition;
  willApply: boolean;
  currentProvenance: FieldProvenance | null;
  incomingSource: string;
}

export interface ImportCandidate {
  id: string;
  source: string;
  identity: GameIdentity;
  action: ImportAction;
  reason: ImportReason;
  matches: IdentityMatch[];
  targetGameId: string | null;
  fieldDiff: FieldDiff[];
  record: ImportSourceRecord;
}

export interface PreviewImportRequest {
  source: string;
  records: ImportSourceRecord[];
}

export interface ImportPreview {
  previewId: string;
  source: string;
  candidates: ImportCandidate[];
  createdAt: string;
  writeCount: number;
}

export interface ImportDecision {
  candidateId: string;
  action: ImportAction;
  targetGameId: string | null;
}

export interface ApplyImportRequest {
  preview: ImportPreview;
  decisions: ImportDecision[];
  idempotencyKey: string;
}

export type ApplyItemStatus =
  | "created"
  | "updated"
  | "merged"
  | "no_changes"
  | "ignored"
  | "conflict"
  | "failed"
  | "already_applied";

export interface ApplyItemResult {
  candidateId: string;
  itemIdempotencyKey: string;
  action: ImportAction;
  status: ApplyItemStatus;
  gameId: string | null;
  message: string;
  appliedFields: string[];
  preservedFields: string[];
}

export interface FieldProvenanceChange {
  itemIdempotencyKey: string;
  gameId: string;
  field: string;
  before: unknown;
  after: unknown;
  previous: FieldProvenance | null;
  current: FieldProvenance;
}

export interface ApplyImportResponse {
  jobId: string;
  idempotencyKey: string;
  replayed: boolean;
  results: ApplyItemResult[];
  provenanceChanges: FieldProvenanceChange[];
}

export type LibraryHealthState = "healthy" | "needs_attention" | "degraded";

export interface LibraryHealthIssue {
  code: string;
  severity: string;
  message: string;
  gameIds: string[];
}

export interface LibraryHealthSnapshot {
  state: LibraryHealthState;
  totalGames: number;
  missingLaunchTargets: number;
  duplicateIdentityGroups: number;
  titleRecallGroups: number;
  unresolvedImportConflicts: number;
  provenanceCoverage: number;
  issues: LibraryHealthIssue[];
}

/** Pure feature boundary. A Tauri adapter can implement this without coupling the store to invoke(). */
export interface LibraryApi {
  preview(request: PreviewImportRequest, signal: AbortSignal): Promise<ImportPreview>;
  apply(request: ApplyImportRequest, signal: AbortSignal): Promise<ApplyImportResponse>;
  health(signal: AbortSignal): Promise<LibraryHealthSnapshot>;
}

File: src-tauri/src/services/library/contracts.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlatformIdentity {
    pub source: String,
    pub id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GameIdentity {
    pub launch_path: Option<String>,
    pub platform_id: Option<PlatformIdentity>,
    pub title_fingerprint: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportSourceRecord {
    pub source_record_id: String,
    pub title: String,
    pub launch_path: Option<String>,
    pub install_dir: Option<String>,
    pub platform_id: Option<PlatformIdentity>,
    pub launch_uri: Option<String>,
    #[serde(default)]
    pub fields: BTreeMap<String, Value>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ImportAction {
    Create,
    Update,
    Merge,
    Conflict,
    Ignore,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ImportReasonCode {
    NewIdentity,
    LaunchPathMatch,
    PlatformIdMatch,
    AmbiguousStrongIdentity,
    TitleRecallOnly,
    NoLaunchTarget,
    ExplicitMerge,
    ExplicitIgnore,
    StalePreview,
    InvalidDecision,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportReason {
    pub code: ImportReasonCode,
    pub message: String,
    #[serde(default)]
    pub recalled_game_ids: Vec<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum IdentityMatchKind {
    PlatformId,
    LaunchPath,
    TitleRecall,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IdentityMatch {
    pub game_id: String,
    pub game_title: String,
    pub kind: IdentityMatchKind,
    pub confidence: f32,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FieldDiffDisposition {
    Unchanged,
    FillEmpty,
    ReplaceImported,
    PreserveUser,
    Conflict,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldProvenance {
    pub game_id: String,
    pub field: String,
    pub source: String,
    pub source_record_id: String,
    pub imported_at: String,
    pub applied_value: Value,
    pub value_hash: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldDiff {
    pub field: String,
    pub current: Value,
    pub incoming: Value,
    pub disposition: FieldDiffDisposition,
    pub will_apply: bool,
    pub current_provenance: Option<FieldProvenance>,
    pub incoming_source: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportCandidate {
    pub id: String,
    pub source: String,
    pub identity: GameIdentity,
    pub action: ImportAction,
    pub reason: ImportReason,
    pub matches: Vec<IdentityMatch>,
    pub target_game_id: Option<String>,
    pub field_diff: Vec<FieldDiff>,
    pub record: ImportSourceRecord,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PreviewImportRequest {
    pub source: String,
    pub records: Vec<ImportSourceRecord>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreview {
    pub preview_id: String,
    pub source: String,
    pub candidates: Vec<ImportCandidate>,
    pub created_at: String,
    pub write_count: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportDecision {
    pub candidate_id: String,
    pub action: ImportAction,
    pub target_game_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyImportRequest {
    pub preview: ImportPreview,
    #[serde(default)]
    pub decisions: Vec<ImportDecision>,
    pub idempotency_key: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ApplyItemStatus {
    Created,
    Updated,
    Merged,
    NoChanges,
    Ignored,
    Conflict,
    Failed,
    AlreadyApplied,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldProvenanceChange {
    pub item_idempotency_key: String,
    pub game_id: String,
    pub field: String,
    pub before: Value,
    pub after: Value,
    pub previous: Option<FieldProvenance>,
    pub current: FieldProvenance,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyItemResult {
    pub candidate_id: String,
    pub item_idempotency_key: String,
    pub action: ImportAction,
    pub status: ApplyItemStatus,
    pub game_id: Option<String>,
    pub message: String,
    #[serde(default)]
    pub applied_fields: Vec<String>,
    #[serde(default)]
    pub preserved_fields: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyImportResponse {
    pub job_id: String,
    pub idempotency_key: String,
    pub replayed: bool,
    pub results: Vec<ApplyItemResult>,
    #[serde(default)]
    pub provenance_changes: Vec<FieldProvenanceChange>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LaunchErrorKind {
    NotFound,
    PermissionDenied,
    InvalidDescriptor,
    UnsupportedScheme,
    SpawnFailed,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum LaunchDescriptor {
    Executable {
        path: String,
        args: Vec<String>,
        working_dir: Option<String>,
    },
    Uri {
        uri: String,
    },
    Unavailable {
        reason: String,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum LaunchOutcome {
    Started {
        descriptor: LaunchDescriptor,
        pid: Option<u32>,
    },
    Delegated {
        descriptor: LaunchDescriptor,
    },
    Failed {
        descriptor: LaunchDescriptor,
        error_kind: LaunchErrorKind,
        message: String,
        retryable: bool,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LibraryHealthState {
    Healthy,
    NeedsAttention,
    Degraded,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LibraryHealthIssue {
    pub code: String,
    pub severity: String,
    pub message: String,
    #[serde(default)]
    pub game_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LibraryHealthSnapshot {
    pub state: LibraryHealthState,
    pub total_games: usize,
    pub missing_launch_targets: usize,
    pub duplicate_identity_groups: usize,
    pub title_recall_groups: usize,
    pub unresolved_import_conflicts: usize,
    pub provenance_coverage: f32,
    pub issues: Vec<LibraryHealthIssue>,
}
